include/zephyr/kernel/internal/sched_priq.h

```diff
@@ -44,15 +48,15 @@ void z_priq_rb_remove(struct _priq_rb *pq, struct k_thread *thread);
 struct k_thread *z_priq_rb_best(struct _priq_rb *pq);
 
 /* Traditional/textbook "multi-queue" structure.  Separate lists for a
- * small number (max 32 here) of fixed priorities.  This corresponds
+ * small number of fixed priorities.  This corresponds
  * to the original Zephyr scheduler.  RAM requirements are
  * comparatively high, but performance is very fast.  Won't work with
  * features like deadline scheduling which need large priority spaces
  * to represent their requirements.
  */
 struct _priq_mq {
-	sys_dlist_t queues[32];
-	unsigned int bitmask; /* bit 1<<i set if queues[i] is non-empty */
+	sys_dlist_t queues[NUM_PRIO];
+	uintptr_t bitmask[PRIQ_BITMAP_SIZE];
```

Comment:
I don't think that `uintptr_t` is the correct type to use here as there does not seem to be anything about the use of `bitmask[]` that suggests that pointers are necessary. My recommendation would be to use a `uint32_t` to make it clear how many bits each element in `bitmask[]` is supposed to have as `uintptr_t` varies in size.

Reply:
I agree with you. Using uintptr_t is for effective. In 64-bit machine we can use 64-bit `clz` operation.
You can followed [here.](https://github.com/zephyrproject-rtos/zephyr/pull/68318#discussion_r1473132254)

Reply:
Likely the optimal solution is whatever the architecture word size is, which would explain the Linux kernel 's BITS_TO_LONGS, etc.

https://elixir.bootlin.com/linux/v4.5/ident/BITS_TO_LONGS

This is one area where we could probably improve. We do have sys/bitarray.h but the issue there is that it pulls in a spinlock.

Reply:
In general, I prefer to use `uint32` as bitarray length even It will lost little performance; and the data type `uint32_t` is enough, more general and sys_bitarray do the same thing.
The reason that I didn't want to use sys_bitarray will pull in spinlock and it's unnecessary.

Reply:
@TaiJuWu

uint32_t is probably safe here. Could you please create an issue for bitarray and reference it (`#1234`) here in a comment?

I'm thinking
* remove spinlock (user should synchronize)
* maybe provide additional macro accessors?

Likely these are things that should probably be fixed.

Reply:
Sure. I will do that recently.

Reply:
> @TaiJuWu
> 
> uint32_t is probably safe here. Could you please create an issue for bitarray and reference it (`#1234`) here in a comment?
> 
> I'm thinking
> 
> * remove spinlock (user should synchronize)
> * maybe provide additional macro accessors?
> 
> Likely these are things that should probably be fixed and the same issues have been mentioned elsewhere.

Sure. I can do that recently.